### PR TITLE
Changed pattern of URL to match appropriate URL's

### PR DESCRIPTION
**Pull request by Shadowmad**
I have changed pattern of the url validator to match correct URLs so that error #10533 would not occure again. As explained in 10533, URLs are passing the tests, but URLs are not correct, this issue was encountered dut to the pattern issues. I ran PHPUnit tests and now everything is passing as well as checked URL in variety of programs to check if URL is correct or not. Please review and let me know if I should re do it or accepted

Thank you!

### CODE REVIEW

**Comment by SilverFire** [comment]
Thank you for your effort.
Could you add a test case for original problem? (http://ya.1ru.!@.!$%^)

File: framework/validators/UrlValidator.php [hunk 1]
```diff
@@ -28,7 +28,7 @@ class UrlValidator extends Validator
      * The pattern may contain a `{schemes}` token that will be replaced
      * by a regular expression which represents the [[validSchemes]].
      */
-    public $pattern = '/^{schemes}:\/\/(([A-Z0-9][A-Z0-9_-]*)(\.[A-Z0-9][A-Z0-9_-]*)+)/i';
+    public $pattern = '/^({schemes}:\\/\\/)(www\\.)?([0-9u0000-uFFFF]*\\.\\-)?([0-9u0000-uFFFF\\-]*)\\.([a-zA-Z]{2,6})?([u0000-uFFFF]{2,6})\\/?([u0000-uFFFF0-9_\\-\\?=\\.\\&])*?/i';
```
Comment (SilverFire) [inline framework/validators/UrlValidator.php]: Why did you add additional `\` everywhere?

Comment (SilverFire) [inline framework/validators/UrlValidator.php]: Why did you add `(www\\.)?` as a separated case?

Comment (SilverFire) [inline framework/validators/UrlValidator.php]: `0-9u0000-uFFFF` what is it? Very strange condition

**Comment by SilverFire** [comment]
The Regexp is overcomplicated and must be simplified.

File: framework/validators/UrlValidator.php (hunk 1 above)
Comment (Shadowmad) [inline framework/validators/UrlValidator.php]: Good evening,

I added www as URL might have www in their name. It is optional, I can
remove that, just thought that would be good addition.
Also, double slash to worked for me in PHP as ultimate regular expression
work. Previously I had troubles with no double slash.
I can create test case for original problem. That is not a problem.

Please let me know if that is good and I do the fixes asap.

Thank you!

Best Regards,
Maksud Aghayev

*Maksud Aghayev *| Full LAMP stack web application developer/PHP Web
Developer
Linkedin: *https://www.linkedin.com/in/maksudaghayev
<https://www.linkedin.com/in/maksudaghayev>*

Comment (SilverFire) [inline framework/validators/UrlValidator.php]: > I added www as URL might have www in their name. It is optional, I can
> remove that, just thought that would be good addition.

The regular expression MUST allow `www.domain.com` as a regular case, 'cause it's the same as `subdomain.domain.com`

Comment (SilverFire) [inline framework/validators/UrlValidator.php]: > Also, double slash to worked for me in PHP as ultimate regular expression
> work. Previously I had troubles with no double slash.

Previously we had no troubles without double backslash. It's not necessary to add it, because it does not fix or enhance anything

Comment (Shadowmad) [inline framework/validators/UrlValidator.php]: I can fix double back-slash issue, that not a problem. Regarding www and subdomain: what if the url goes as follows www.subdomain.domain.com? Will regular expression still work? With the one I wrote it works by checking www and subdomain. Could you clarify that please

Comment (Shadowmad) [inline framework/validators/UrlValidator.php]: Also, u0000-uFFFF is all possible Unicode. I can change that to 0-9a-zA-Z to limit regular expression to only English words.

Comment (SilverFire) [inline framework/validators/UrlValidator.php]: > www.subdomain.domain.com? Will regular expression still work

yes, it's a valid URL

Comment (SilverFire) [inline framework/validators/UrlValidator.php]: We need to respect `enableIDN` option

Comment (samdark) [inline framework/validators/UrlValidator.php]: Unicode ranges could be replaced with `\pL`

**Comment by cebe** [comment]
The regex itself should not allow UTF8 characters, utf8 is handled by converting the IDN name first and then checking against the pattern.

**Comment by samdark** [comment]
@cebe is right about it.

**Comment by samdark** [comment]
I propose fixing it by adding `(:?$|/)` to the end of existing regex.

**Comment by Shadowmad** [comment]
@samdark That might help with escaping issues of not recognized symbols, byt what about patterns like https://www.google.de/search?q=yii+framework&ie=utf-8&oe=utf-8.&rls=org.mozilla:de:official&client=firefox-a&gws_rd=cr

How do you work them around? Trough regexp or trough some system in the code? I apologize in advance, I am still trying to figure out the code and dependencies in the whole system

I came up with something like this, taking in cosideration what you had and suggestions of @samdark 
^(http|https):\/\/(([A-Z0-9][A-Z0-9_-]*)(\.[A-Z0-9][A-Z0-9_-]*)+)([\.\/\?\w\d=\_\-&\%\+\:])*

Could you please provide feedback on that and I would update my PR as well as add test cases as requested above. 

Thank you

**Comment by SilverFire** [comment]
> [\.\/\?\w\d=\_\-&\%\+\:]

I don't like this part. All we need to do - to stop matching _domain name_ part of the string. We mustn't take care anything else in the URL.

The original problem was a domain `http://ya.1ru.!@.!$%^`, that is not a valid domain, for sure, but `http://ya.1ru/!@.!$%^` might be a valid one.

I'm suggesting the following regexp:

`/^(http|https):\/\/((?:[A-Z0-9][A-Z0-9_-]*)(?:\.[A-Z0-9][A-Z0-9_-]*)+)(?::?|\/|\?)/i`

`(?::?|\/|\?)` means stop matching domain name on `:`, `/` or `?` character.

**Comment by Shadowmad** [comment]
I apologize for inconvenience. I totally misunderstood the problem. If I
need to cut out URL as only to domain name part then proposed fix is better
one, I thought that I need to check for possible folder names, variables or
"pretty" URLs. I will take that into consideration tomorrow and apply all
the fixed if that is still applicable.

Thank you for guidance!

Best Regards,
Maksud Aghayev

*Maksud Aghayev *| Full LAMP stack web application developer/PHP Web
Developer
Linkedin: *https://www.linkedin.com/in/maksudaghayev
<https://www.linkedin.com/in/maksudaghayev>*

**Comment by SilverFire** [comment]
Closed in favor or #10753
